README.md

```diff
@@ -200,6 +200,30 @@ python3 -m odxtools list -a "$YOUR_PDX_FILE"
   # -> decoded response: [session()]
   ```
 
+## Using the non-strict mode
+
+By default, odxtools raises exceptions if it suspects that it cannot
+fulfill a requested operation correctly. For example, if the dataset
+it is instructed to load is detected to be not conformant with the ODX
+specification, or if completing the operation requires missing
+features of odxtools. To be able to deal with such cases, odxtools
+provides a "non-strict" mode where such issues are ignored, but where
+the results are undefined. The following snippet shows how to instruct
+odxtools to load a non-conforming file in non-strict mode, and after
+this is done, enables the safety checks again:
+
+  ```python
+  import odxtools
+
+  [...]
+
+  odxtools.exceptions.strict_mode = False
+  botched_db = odxtools.load_file("my_non-conforming_database.pdx")
+  odxtools.exceptions.strict_mode = True
```

Comment:
To avoid issues with threads and changing global variables side effect, it would be better to either 
* Make `strict_mode` an argument to `load_file`
* Use https://docs.python.org/3/library/warnings.html with default error for OdxWarning, and users can change behavior using https://docs.python.org/3/library/warnings.html#warning-filter

Reply:
I considered this, but that would mean to drag the `strict` argument through to all constructors, `from_et()` functions and `odxraise()`, `odxassert()` and `odxrequire()` making them *very* clunky IMO. Also, I came to the conclusion that using one thread in strict mode and another in non-strict mode is quite a corner case...

> Use https://docs.python.org/3/library/warnings.html with default error for OdxWarning, and users can change behavior using https://docs.python.org/3/library/warnings.html#warning-filter

does this work with type checking? (making mypy happy is the reason for most of these assertations...)

Reply:
* Yes it works with type checking
* I left you a few review comments on how it should look like in `odxtools/exceptions.py`, the only change from your current code would be

Usage for end user
```py
  with warnings.filterwarnings("ignore", OdxError): # Could be "once", "default", "always"
      botched_db = odxtools.load_file("my_non-conforming_database.pdx")
```

Reply:
ok, I'll have a look. BTW: this also seems to basically be a global variable to me?

Reply:
I now have an (almost) working version of this, but I've got very mixed feelings: On one hand, it feels more standard to do this using the `warnings` module, on the other hand, it is much more brittle because the unit testing modules (`unittest` and `pytest`) seem to regard the `warnings` module as their personal property and thus mess with it pretty freely. In particular `unittest`: It modifies `warnings`'s filters *by default*, and one explicitly has to  tell it to leave `warnings` alone (i.e., `unittest.main(warnings=???)`. unfortunately, I haven't yet figured out what it expects for the `warnings` argument). `pytest` is a bit less offensive as it only messes with the filters if its config says so. (ours currently does which confused me for quite a while.)

@kayoub5: Do you think going the `warnings` route is worth it  anyway? if yes, do you know how to convince `unittests` to leave the filters of `warnings` alone?

Reply:
> Can be provided in the unittest.main https://python.readthedocs.io/en/stable/library/unittest.html#unittest.main

yes, but what must it be set to? setting `warning="error"` will break some tests because some currently non-fatal warnings will become raised exceptions, whilst setting it to "ignore" (or not specifying it at all) makes those tests fail which expect an `OdxError` to be raised. (The `unittest` module seems to always set some catch-all warning filters and it seems like it cannot be instructed to leave these filters alone...)

tl;dr: I think we should stick with a "custom" global variable instead of trying to use that of the `warnings` module, at least for now...

Reply:
You can specify per type `error::OdxError`

See https://docs.python.org/3/library/warnings.html#default-warning-filter

Reply:
yes, but not with `unittest` (at least not without some severe hack), because `unittest` always adds some catch-all filters to the beginning of the list...

Problems like this are probably the reason why global variables ought to be avoided (i.e., multiple separate parts of the code fighting for supremacy over it)...

Reply:
Rest of the code looks OK, if you think that using `warnings` is not worth the effort, then LGTM

Reply:
> yes, but not with unittest (at least not without some severe hack), because unittest always adds some catch-all filters to the beginning of the list...

That's one of the many reasons people abandoned unittest and switched to pytest

Reply:
okay, let's get this merged then. Note that the same issue can happen with user code that using odxtools: if it modifies the filters of `warnings`, and if we relied `warnings` for controlling odxtools' behavior, the side effects could be very non-intuitive.

if you want to have a look at this issue yourself, check out [this branch](https://github.com/andlaus/odxtools/tree/WIP)...